### PR TITLE
Improve calibration error speed by replacing `for` loop

### LINKED ISSUE #767: Remove for loop on calibration error.
## 🐛 Improvement

More than a bug I wanted to see if the calibration error could be speed up by removing the loops and using native pytorch operations. Please correct me if I am wrong but the current implementation could go from:

```python
conf_bin = torch.zeros_like(bin_boundaries)
acc_bin = torch.zeros_like(bin_boundaries)
prop_bin = torch.zeros_like(bin_boundaries)
for i, (bin_lower, bin_upper) in enumerate(zip(bin_boundaries[:-1], bin_boundaries[1:])):
    # Calculated confidence and accuracy in each bin
    in_bin = confidences.gt(bin_lower.item()) * confidences.le(bin_upper.item())
    prop_in_bin = in_bin.float().mean()
    if prop_in_bin.item() > 0:
        acc_bin[i] = accuracies[in_bin].float().mean()
        conf_bin[i] = confidences[in_bin].mean()
        prop_bin[i] = prop_in_bin
```
To something similar to this:
```python
acc_bin = torch.zeros(len(bin_boundaries)-1)
conf_bin = torch.zeros(len(bin_boundaries)-1)
count_bin = torch.zeros(len(bin_boundaries)-1)

# Compute the position of each confidence in the boundaries.
indices = torch.bucketize(confidences, bin_boundaries) - 1

# Compute count per bin.
count_bin.scatter_add_(dim=0, index=indices, src=torch.ones_like(confidences))

# Add up each confidence on the respective bin and get mean.
conf_bin.scatter_add_(dim=0, index=indices, src=confidences)
conf_bin = torch.nan_to_num(conf_bin / count_bin)

# Add up number of correct points on each bin and get accuracy.
acc_bin.scatter_add_(dim=0, index=indices, src=accuracies)
acc_bin = torch.nan_to_num(acc_bin / count_bin)

# Go from counts to proportions.
prop_bin = count_bin / count_bin.sum()
```

### PR DESCRIPTION
**Pull request by ramonemiliani93**
## What does this PR do?

Improve calibration error speed by removing for loop and using `bucketize` + `scatter_add`.
Removes the for loop in the calibration error and uses bucketize with scatter add for improved speed (~10x).

Fixes #767

## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Oh yes 😎 

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
Nice, have you measured the perfomance? :rabbit:

**Comment by ramonemiliani93** [comment]
@Borda Yes 👌 Only on CPU though, here's the script:
```python
import timeit

import torch


def method_a(confidences, accuracies, bin_boundaries):
    def _method_a():
        conf_bin = torch.zeros_like(bin_boundaries)
        acc_bin = torch.zeros_like(bin_boundaries)
        prop_bin = torch.zeros_like(bin_boundaries)
        for i, (bin_lower, bin_upper) in enumerate(
            zip(bin_boundaries[:-1], bin_boundaries[1:])
        ):
            # Calculated confidence and accuracy in each bin
            in_bin = confidences.gt(bin_lower.item()) * confidences.le(bin_upper.item())
            prop_in_bin = in_bin.float().mean()
            if prop_in_bin.item() > 0:
                acc_bin[i] = accuracies[in_bin].float().mean()
                conf_bin[i] = confidences[in_bin].mean()
                prop_bin[i] = prop_in_bin

    return _method_a


def method_b(confidences, accuracies, bin_boundaries):
    def _method_b():
        acc_bin = torch.zeros(len(bin_boundaries) - 1)
        conf_bin = torch.zeros(len(bin_boundaries) - 1)
        count_bin = torch.zeros(len(bin_boundaries) - 1)

        indices = torch.bucketize(confidences, bin_boundaries) - 1

        count_bin.scatter_add_(dim=0, index=indices, src=torch.ones_like(confidences))

        conf_bin.scatter_add_(dim=0, index=indices, src=confidences)
        conf_bin = torch.nan_to_num(conf_bin / count_bin)

        acc_bin.scatter_add_(dim=0, index=indices, src=accuracies)
        acc_bin = torch.nan_to_num(acc_bin / count_bin)

        prop_bin = count_bin / count_bin.sum()

    return _method_b


n_bins = 20
size = (10000000,)
confidences = torch.rand(size)
accuracies = torch.randint(low=0, high=2, size=size).float()
bin_boundaries = torch.linspace(0, 1, steps=n_bins + 1)

t = timeit.Timer(method_a(confidences, accuracies, bin_boundaries))
print(t.timeit(100))

t = timeit.Timer(method_b(confidences, accuracies, bin_boundaries))
print(t.timeit(100))
```
The time depends on the `size` and `n_bins`, for the values I set the difference was from 60 s to 11.4 s on 100 runs which speeds it up ~6x.

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
seems that older PyTorch versions do not have `bucketize`
```
AttributeError: module 'torch' has no attribute 'bucketize'
```

**Comment by ramonemiliani93** [comment]
Should a `try-except` block do it?
```python
    acc_bin = torch.zeros(len(bin_boundaries) - 1, device=confidences.device)
    conf_bin = torch.zeros(len(bin_boundaries) - 1, device=confidences.device)
    count_bin = torch.zeros(len(bin_boundaries) - 1, device=confidences.device)
    prop_bin = torch.zeros(len(bin_boundaries) - 1, device=confidences.device)
    
    try:
        indices = torch.bucketize(confidences, bin_boundaries) - 1

    except AttributeError:
        for i, (bin_lower, bin_upper) in enumerate(zip(bin_boundaries[:-1], bin_boundaries[1:])):
            # Calculated confidence and accuracy in each bin
            in_bin = confidences.gt(bin_lower.item()) * confidences.le(bin_upper.item())
            prop_in_bin = in_bin.float().mean()
            if prop_in_bin.item() > 0:
                acc_bin[i] = accuracies[in_bin].float().mean()
                conf_bin[i] = confidences[in_bin].mean()
                prop_bin[i] = prop_in_bin
    else:
        count_bin.scatter_add_(dim=0, index=indices, src=torch.ones_like(confidences))
        conf_bin.scatter_add_(dim=0, index=indices, src=confidences)
        conf_bin = torch.nan_to_num(conf_bin / count_bin)
        acc_bin.scatter_add_(dim=0, index=indices, src=accuracies)
        acc_bin = torch.nan_to_num(acc_bin / count_bin)
        prop_bin = count_bin / count_bin.sum()
```

**Comment by Borda** [comment]
> Should a `try-except` block do it?

I would rather make it explicit so define the blocked as protected functions:
```py
def _loop_block(...):
    for i, (bin_lower, bin_upper) in enumerate(zip(bin_boundaries[:-1], bin_boundaries[1:])):
        # Calculated confidence and accuracy in each bin
        in_bin = confidences.gt(bin_lower.item()) * confidences.le(bin_upper.item())
        prop_in_bin = in_bin.float().mean()
        if prop_in_bin.item() > 0:
            acc_bin[i] = accuracies[in_bin].float().mean()
            conf_bin[i] = confidences[in_bin].mean()
            prop_bin[i] = prop_in_bin
    return ...

def _enrolled_block(...):
    indices = torch.bucketize(confidences, bin_boundaries) - 1
    count_bin.scatter_add_(dim=0, index=indices, src=torch.ones_like(confidences))
    conf_bin.scatter_add_(dim=0, index=indices, src=confidences)
    conf_bin = torch.nan_to_num(conf_bin / count_bin)
    acc_bin.scatter_add_(dim=0, index=indices, src=accuracies)
    acc_bin = torch.nan_to_num(acc_bin / count_bin)
    prop_bin = count_bin / count_bin.sum()
    return ...
```
then func names are just illustrations...
and later integrate as:
```py
    if hasattrib(torch, "bucketize"):
        _enrolled_block(...)
    else:
        _loop_block(...)
```

**Comment by SkafteNicki** [comment]
> seems that older PyTorch versions do not have `bucketize` it seems it was added in 1.8
> 
> ```
> AttributeError: module 'torch' has no attribute 'bucketize'
> ```
> 
> said so we can have a hard switch if needed - older versions will use loop, new version this fancy solutions :]

@borda sure about this?
We have the following code elsewhere also relying on `bucketize`:
https://github.com/PyTorchLightning/metrics/blob/c519402693a6cb235367614f192d67332cbb4bc0/torchmetrics/functional/classification/auroc.py#L105-L109
From that it seems that 1.6 was the point it was introduced.

**Comment by Borda** [comment]
> We have the following code elsewhere also relying on `bucketize`:

good point, I did not dive into details, just saw that all tests lower 1.8 was failing (could be also other reason lol)

**Comment by Borda** [comment]
@ramonemiliani93 do you think we can finish it so we can include it in the next bug-fix release? :rabbit:

**Comment by SkafteNicki** [comment]
@Borda should be taken care of now

File: torchmetrics/functional/classification/calibration_error.py [hunk 1]
```diff
@@ -18,6 +18,65 @@
...
+        confidences (FloatTensor): The confidence (i.e. predicted prob) of the top1 prediction.
+        accuracies (FloatTensor): 1.0 if the top-1 prediction was correct, 0.0 otherwise.
+        bin_boundaries (FloatTensor): Bin boundaries separating the linspace from 0 to 1.
+
+    Returns:
+        tuple with binned accuracy, binned confidence and binned probabilities
+    """
+    acc_bin = torch.zeros(len(bin_boundaries) - 1, device=confidences.device)
+    conf_bin = torch.zeros(len(bin_boundaries) - 1, device=confidences.device)
+    count_bin = torch.zeros(len(bin_boundaries) - 1, device=confidences.device)
```
Comment (justusschock) [inline torchmetrics/functional/classification/calibration_error.py:66]:
```suggestion
    acc_bin = torch.zeros(len(bin_boundaries) - 1, device=confidences.device, dtype=confidences.dtype)
    conf_bin = torch.zeros(len(bin_boundaries) - 1, device=confidences.device, dtype=confidences.dtype)
    count_bin = torch.zeros(len(bin_boundaries) - 1, device=confidences.device, dtype=confidences.dtype)
```

**Review by justusschock — APPROVED** [review]

**Comment by ramonemiliani93** [comment]
@Borda Sorry for the late reply! I have had a lot of things on my side 😔 I just saw that @SkafteNicki already solved it. I hope to be able to contribute more next time!

**Comment by SkafteNicki** [comment]
It complains about `torch.nan_to_num` which was introduced in 1.8...
I increase the required version for the faster version :]

**Pull request merged**